mirage/tar_mirage.ml

```diff
@@ -427,9 +455,119 @@ module Make_KV_RW (CLOCK : Mirage_clock.PCLOCK) (BLOCK : Mirage_block.S) = struc
+        in
+        write t (div start_bytes sector_size) data')
 
-  let allocate _ _ ?last_modified:_ _ =
-    Lwt.return (Error `Append_only)
+  let allocate t key ?last_modified size =
+    Lwt_mutex.with_lock t.write_lock (fun () ->
+        let ( >>>= ) = Lwt_result.bind in
+        let r =
+          let ( let* ) = Result.bind in
+          (* XXX: map `Entry_already_exists to `Append_only ?! *)
+          let* () = is_safe_to_set t key in
+          let hdr = header_of_key ?last_modified key (Optint.Int63.to_int size) in
+          let space_needed = space_needed hdr in
+          let* () = if free t >= space_needed then Ok () else Error `No_space in
+          Ok (hdr, space_needed)
+        in
+        Lwt.return r >>>= fun (hdr, space_needed) ->
+        let open Int64 in
+        let sector_size = of_int t.info.Mirage_block.sector_size in
+        let header_start_bytes =
+          sub t.end_of_archive (of_int (2 * Tar.Header.length))
+        in
+        (* we don't have to zero the block before end_of_archive as it is
+           already zero due to the terminating double zero block (sentinel) *)
+        let to_zero_start_bytes = t.end_of_archive in
+        (* space_needed - 1 block + sentinel = space_needed + 1 block *)
+        let end_bytes = add t.end_of_archive
+            (add space_needed (of_int Tar.Header.length)) in
+        (* Compute the starting sector and ending sector *)
+        let to_zero_start_sector, to_zero_start_sector_offset =
+          div to_zero_start_bytes sector_size,
+          rem to_zero_start_bytes sector_size
+        in
+        let last_sector_offset = rem end_bytes sector_size in
+        let end_sector = div (add end_bytes sector_size) sector_size in
+        (* [num_to_zero_sectors] is at least 1 as we need to write at least one
+           zero block of the new sentinel. *)
+        let num_to_zero_sectors = to_int (sub end_sector to_zero_start_sector) in
+        let zero_sector = Cstruct.create t.info.Mirage_block.sector_size in
+        let data = Array.init num_to_zero_sectors (fun _ -> zero_sector) in
+        let nonzero_sector c =
+          (* we allocate a new buffer if [c] is [zero_sector], otherwise we can
+             reuse it in the case first and last sectors are the same. *)
+          if c != zero_sector then c else Cstruct.create (Cstruct.length c)
+        in
+        (* Read slack at start and end sector(s) *)
+        let () = data.(0) <- nonzero_sector data.(0) in
+        read_partial_sector t to_zero_start_sector data.(0)
+          ~offset:0L
+          ~length:to_zero_start_sector_offset
+        >>>= fun () ->
+        let last = nonzero_sector data.(num_to_zero_sectors - 1) in
+        let () = data.(num_to_zero_sectors - 1 ) <- last in
+        read_partial_sector t (pred end_sector) last
+          ~offset:last_sector_offset
+          ~length:(sub sector_size last_sector_offset) >>>= fun () ->
```

Comment:
We may need to blit in data from the beginning and the end. We try to avoid an extra allocation in the case the first and the last sector is the same sector.